### ISSUE TITLE
Implement bulk data upload views

### ISSUE THREAD

**Issue by azerbini** [azerbini/eamena_dev#27]
This issue elaborates on:  https://github.com/azerbini/eamena_v3/issues/80

The desired outcome is to allow users to test bulk upload spreadsheets (testing is handled by this class: https://github.com/azerbini/eamena_dev/blob/eamena-migration-branch/arches/management/commands/Excelreader.py) and bulk load data without having to learn how to use shell commands.

Features:

1) A link to a bulk data load view accessible from the Resource Manager
2) A view that would contain the following features: a file upload box to upload xlsx spreadsheet files; an html log of the tests performed on the uploaded spreadsheet. When all tests are passed, an arches file is generated in the background and the user is given the option to load the data straight into the database; once upload is successful, an html log listing the LOADID of the uploaded file.

**Comment by mradamcox** [comment]
@azerbini I'm working on this but my first test of the Excelreader class failed with an exception, and I see now that there is no logging in that class, only exceptions. I thought that the desired workflow was for a user to test a spreadsheet and get a read out of all errors in it so these can be corrected. However, I'm not sure how to do that if the Excelreader command fails and exits when it encounters an error.

To test this, I just downloaded the template spreadsheet that you passed along for E27 resources. I ran

    python manage.py Excelreader -o site_dataset -s "C:\arches\eamena\builk_upload testing\E27  Heritage Place Template.xlsx" -r HERITAGE_RESOURCE_GROUP.E27

and I got this error:

```
Traceback (most recent call last):
  File "manage.py", line 28, in <module>
    execute_from_command_line(sys.argv)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 33, in handle
    self.SiteDataset(options['source'], options['resource_type'], options['dest_dir'],options['append_data'])
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 175, in SiteDataset
    self.validate_rows_and_values(wb2)
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 84, in validate_rows_and_values
    raise ValueError("Error: cells in sheet %s do not contain an equal number of semicolon separated values or are empty. Errors are at the following lines: %s" % (workbook.sheetnames[sheet_index], sorted(ret)))
ValueError: Error: cells in sheet zzAssessment do not contain an equal number of semicolon separated values or are empty. Errors are at the following lines: [2, 3, 4, 5, 6]
```

It seems like the script has failed after testing only the `zzAssessment` sheet, though there are many more in the workbook.

I'm not sure how to use this command in terms of the iterative testing you outlined. However, I'm also not sure that I'm using the management command correctly, so let me know when you can...

**Comment by azerbini** [comment]
You are using the command correctly, @mradamcox . Excelreader performs various checks on the spreadsheet:
1) checks the headings exist as db nodes
2) checks the number of rows is the same across all tabs
3) checks the number of semicolon separated values is the same in each row of each tab except for the NOT tab
4) checks that there are no empty cells in the sequence.

Once these first checks are done, the evaluation of the individual cells begins, where the script will check for:
i) Dates formatted in the right way
ii) Geometries formatted in a WKT compatible way
iii) Concepts in a given column correspond to concepts existing within the node of the name of the heading of that column


It would appear your spreadsheet is failing tests 3-4 on the zzAssessment tab.

I never got round to consolidating all the errors into a comprehensive log. This could be done as part of this ticket.



Sent from my iPhone

On 29 May 2018, at 22:01, mradamcox <notifications@github.com<mailto:notifications@github.com>> wrote:


@azerbini<https://github.com/azerbini> I'm working on this but my first test of the Excelreader class failed with an exception, and I see now that there is no logging in that class, only exceptions. I thought that the desired workflow was for a user to test a spreadsheet and get a read out of all errors in it so these can be corrected. However, I'm not sure how to do that if the Excelreader command fails and exits when it encounters an error.

To test this, I just downloaded the template spreadsheet that you passed along for E27 resources. I ran

python manage.py Excelreader -o site_dataset -s "C:\arches\eamena\builk_upload testing\E27  Heritage Place Template.xlsx" -r HERITAGE_RESOURCE_GROUP.E27


and I got this error:

Traceback (most recent call last):
  File "manage.py", line 28, in <module>
    execute_from_command_line(sys.argv)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "C:\arches\eamena\dev\env\lib\site-packages\django\core\management\base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 33, in handle
    self.SiteDataset(options['source'], options['resource_type'], options['dest_dir'],options['append_data'])
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 175, in SiteDataset
    self.validate_rows_and_values(wb2)
  File "c:\arches\eamena\dev\arches\arches\management\commands\Excelreader.py", line 84, in validate_rows_and_values
    raise ValueError("Error: cells in sheet %s do not contain an equal number of semicolon separated values or are empty. Errors are at the following lines: %s" % (workbook.sheetnames[sheet_index], sorted(ret)))
ValueError: Error: cells in sheet zzAssessment do not contain an equal number of semicolon separated values or are empty. Errors are at the following lines: [2, 3, 4, 5, 6]


It seems like the script has failed after testing only the zzAssessment sheet, though there are many more in the workbook.

I'm not sure how to use this command in terms of the iterative testing you outlined. However, I'm also not sure that I'm using the management command correctly, so let me know when you can...

**Comment by mradamcox** [comment]
Ok, in that case, to get this working I will need to do quite a bit of work on the Excelreader command. As it is, there is no output from that command that could be passed to a view as described in this ticket.

**Comment by azerbini** [comment]
Sure, that sounds good. I can also jump in on this since I have written Excelreader from scratch and can easily work in the logging. I mean, all you need really is to store the errors created by the various exceptions into a dictionary and then return that to the view for parsing in Javascript.

________________________________

**Comment by mradamcox** [comment]
Yeah, I've done it by adding a new logger using `logging`. I'll push just those changes ASAP so you can see them. I think using the logger like this will be the easiest way to handle this. Then the view can access the same log after the command has been run and push its contents whereever they are needed.

**Comment by mradamcox** [comment]
Ok I have added a new branch here called ` EAMENA-27-bulk-upload-view`, and the only commit I have put on it so far holds the changes I made to create a logger and write the error messages to it that you have put in the exceptions. If you can improve the print out and put that into the logger as well, that would be great. I'm getting the logger in the new view, and that's how its contents can be displayed in the template.

**Comment by mradamcox** [comment]
I think it would be best if you can just put new commits on that branch for now (without merging it in to master).

**Comment by mradamcox** [comment]
@azerbini the above commits add the new page and all functionality up to the point where we need more information from the Excelreader command before moving on.

So far I have just been testing with the template spreadsheet that you passed along for E27 resources.

Specifically, these are what I see as next steps:

1. The SiteDataset() method needs to return a more unified set of messages, as well as a simple pass/fail flag. The portion of that method that makes the .arches file should only be run if all the tests pass (which I realize was the actual outcome originally, when any errors resulted in raised exceptions, but now that error messages are used instead of raising exceptions, we'll need to rework this).
2. The way we pass information from SiteDataset to the view should be reworked. I had set things up to use the built-in logging capabilities, and while this works alright, I think it could run into problems with concurrent use. In the code you'll see I've outlined and mocked up an alternative method in which we can write the validation output to a file handle() and then pass the path of that file to the view where it can be loaded. Unfortunately, passing raw json from the command to the view is not as easy as I was hoping.
3. Once an "all tests pass" flag has been implemented, we can continue with the addition of the final view and ajax call which will load the .arches file itself.
4. We'll have to modify the .arches import command to write the bulk data load id (etc) to a single log file.
5. Finally, we'll be able to add the contents of that log file to the build upload page in a box similar to how I am displaying the excel-reader log contents at present. I'm thinking another ajax call for this.

**Comment by azerbini** [comment]
@mradamcox Working on this now. One other element to add here is the SiteDataset() flag `append` which the user should be allowed to pass from the view, right after passing the Resource Type. So I would add a step after your 2. Choose Resource Type called 3. Upload Type which would have two options: 1) New data 2) Append to existing.

**Comment by mradamcox** [comment]
Ok great. What would you say about combining it with 2? Then the header
could be something like "configure import". This would keep the number of
steps down which I always think is helpful.

**Comment by azerbini** [comment]
Yep great idea!

**Comment by mradamcox** [comment]
Cool, I'll add this when I pick the ticket back up, after some of the Excelreader additions are made as described above.

**Comment by azerbini** [comment]
Okay @mradamcox  I have done a first iteration of the logging on Excelreader (see this commit : 4d97a3d). I have done this in haste and haven't tested this yet at all but perhaps this will give you enough to get on with the view.

**Comment by mradamcox** [comment]
We should also fix the handling of file paths so that only the file name is ever passed via ajax, not the full file path. The BULK_UPLOAD_DIR setting can be used in views.py to complete the full path.

**Comment by mradamcox** [comment]
@azerbini Please see the new commits I have put on the `EAMENA-27-bulk-upload-view` branch. The entire workflow is fully functional for me locally. So far I have only tested it with a subset of one of the spreadsheets you sent me (below), as the larger files took far too long to validate and load.

[E36N34-12-trunc5.xlsx](https://github.com/azerbini/eamena_dev/files/2100122/E36N34-12-trunc5.xlsx)

- Changed the location of the link so it is now an additional entry on the Resource Manager dropdown
- Improved the template inheritance and structure so the user begins at the summary page showing all of the past loads and has a link to the new upload page
- Improved some of the styling and font colors for messages
- Removed all of the old logging remnants
- Finished the views as necessary

**Comment by azerbini** [comment]
Excellent, I will try to look into this tonight. Given how long it takes to test and load bulk data files, I was wondering whether we could include a status bar in the view? How hard do you think it would be?


________________________________

**Comment by mradamcox** [comment]
@azerbini I believe that would require some pretty serious retooling of those functions (the Excelreader and load command). We would probably have to add a component to each one that does a quick analysis and estimates the length of the import process, and then reports that back to the javascript before actually beginning.

**Comment by azerbini** [comment]
025b0ff adds a couple of bug fixes to this. I am getting ready to merge this branch into the main migration branch. However, @mradamcox , I am wondering if we could do one last extension of this code to try and include streaming of the validate_concept test. I am looking into adding an xhr callback to the ajax call and then implement something like this (though of course with a python backend): https://www.binarytides.com/ajax-based-streaming-without-polling/

What do you think?

**Comment by mradamcox** [comment]
Hey @azerbini , I've been looking around and I actually think we would do well to integrate Celery into the app to handle this. It's the most popular python task-queuing library (as far as I know), and there are lots of examples of integrations with django. For some starting points: https://www.dangtrinh.com/2013/07/django-celery-display-progress-bar-of.html, https://stackoverflow.com/questions/7380373/django-celery-progress-bar#15117420 and especially https://buildwithdjango.com/blog/post/celery-progress-bars/. Hopefully I will get a chance to look into this next week.

**Comment by mradamcox** [comment]
Alright, I'd like to wait on the Celery integration until @TeriForey 's PR #37 is merged, as it will likely sit on top of some of the things she has been working on.

**Comment by azerbini** [comment]
Yes, agreed. We do need streaming for this since it just takes far too long for some of these operations to complete and the user is just likely to think an error has occurred unless a progress is given.

Adam, I think it would be useful for the validate_concepts test to report back every time a new node is being evaluated (basically output to the front end the command that currently prints headers to the console). For the actual Arches upload, a more traditional progress bar would work.

**Comment by mradamcox** [comment]
Ok, sounds good. I think that once we're able to get a basic example [like this](https://djangosnippets.org/snippets/2898/) working, we'll be able to do everything we need. I do expect that we'll need to modify the Excelreader command a bit.

**Comment by mradamcox** [comment]
@azerbini and @TeriForey, after finally getting to working more with celery last week and today, I've encountered a few issues:

First, celery requires third-party software on the server to act as a message broker, something like RabbitMQ. While this can be setup and installed pretty easily, it does put more responsibility on the sys admin, and would increase the effort needed to deploy a new installation.

Also, the latest celery release, 4.2, doesn't work with django < 1.8, and we are currently using 1.6. I did a little testing on upgrading django to 1.8, but ran into lots of issues right away, so that wouldn't be a quick effort. Using celery 3.1 is an option, though it is out-of-date.

Finally, celery is not supported on windows (which I discovered on [pypi](https://pypi.org/project/celery/)), which may make developing with celery more difficult, for me at least.

With these things in mind I wanted to take a step back and re-evaluate what we really need: To improve the reporting of the validate and import commands when users are doing this via the new bulk loader. When weighing this I'm now more inclined to go the route of splitting the validation command into smaller pieces, and chaining together more sequential ajax calls. At any rate, I wanted to check in with you two before proceeding further.

**Comment by azerbini** [comment]
@mradamcox , whilst I am on board with the idea of breaking up the validation, I am not sure how you'd go about it, particularly when it comes to validating concept values. Also, we would need to break up the actual data upload too, since that too can be a very lengthy process if many records (or highly enhanced records, or a combination of the two) are being loaded.

Perhaps one viable option would be for the script to break the XLSX file into multiple parts (each 100 records or so?), each of which creating a separate bulk upload file. If an error is returned to the user, the partial XLSX files are deleted and the process starts again upon upload of the modified source XLSX file.

**Comment by mradamcox** [comment]
Ok, I'm going to go down this route then, beginning with splitting up the validation functions as best I can.

**Comment by azerbini** [comment]
I am now closing this issue. I will be opening a new ticket requesting an enhancement on the bulk upload interface regarding breaking up the load_resources operation.
